### PR TITLE
fix: etcd watch失败时，内存泄漏 #2919

### LINKED ISSUE #2919: etcd watch失败时，内存泄漏
### Issue description

image
当etcd watch失败，goroutine退出时，会在defer func中将当前GenericStore加入到storeNeedReInit中
image
每两分钟会调用一次ReInit，重新将storeNeedReInit slice中的GenericStore初始化，又回调回到上面出问题的代码，再次加入storeNeedReInit
storeNeedReInit并没有清理的地方，会一直增大

### Expected behavior

就算etcd异常，manager-api进程内存使用也不能持续增长

### How to Reproduce

1.将etcd打断点断住，模拟etcd返回超时
2.反复在dashboard中进行操作
3.放开etcd断点
4.manager-api进程内存使用持续增长

### Screenshots

_No response_

### Environment

- apisix version (cmd: `apisix version`):
- OS (cmd: `uname -a`):
- OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`):
- etcd version, if have (cmd: run `etcd --version`):
- apisix-dashboard version, if have:
- Browser version, if have:

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by hauerwu**
Please answer these questions before submitting a pull request, **or your PR will get closed**.

**Why submit this pull request?**

- [x] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [ ] Backport patches

**What changes will this PR take into?**

Please update this section with detailed description.

**Related issues**

fix/resolve #2919

**Checklist:**

- [x ] Did you explain what problem does this PR solve? Or what new features have been added?
- [ ] Have you added corresponding test cases?
- [ ] Have you modified the corresponding document?
- [ ] Is this PR backward compatible? If it is not backward compatible, please discuss on the mailing list first

### CODE REVIEW

File: api/internal/core/store/store.go [hunk 1]
```diff
@@ -109,12 +110,20 @@ func ReInit() error {
 			return err
 		}
 	}
+
+	//clear storeNeedReInit
+	storeNeedReInit = storeNeedReInit[:0]
```
Comment (hanqingwu) [inline api/internal/core/store/store.go:115]: I think only add this line is enough.

Comment (hauerwu) [reply api/internal/core/store/store.go:115]: when for loop return err, this line will not be excute,and the same store may be add to the storeNeedReInit slice multiply times, and in some condition the storeNeedReInit slice may increase to a very huge size

Comment (hauerwu) [reply api/internal/core/store/store.go:115]: image
as this pic show, manager-api process memory increase to 21GB!

Comment (hanqingwu) [reply api/internal/core/store/store.go:115]: but if return err ,  it can't reach code  
`storeNeedReInit = append(storeNeedReInit, s) `

Comment (hauerwu) [reply api/internal/core/store/store.go:115]: so i add some other protect:

image

Comment (hauerwu) [reply api/internal/core/store/store.go:115]: only add the store to the slice when it's inited changed from true to false

Comment (hanqingwu) [reply api/internal/core/store/store.go:115]: so  if return err , it can't reach code  ,   why multi  times ?

**Review by monkeyDluffy6017 — APPROVED** [review]

**Comment by monkeyDluffy6017** [comment]
@nic-chen @bzp2010 please help to check

File: api/internal/core/store/store.go (hunk 1 above)
Comment (monkeyDluffy6017) [reply api/internal/core/store/store.go:115]: @hanqingwu do you accept this pr?

Comment (hanqingwu) [reply api/internal/core/store/store.go:115]: yes, I think this PR works .

**Review by hanqingwu — APPROVED** [review]

**Review by nic-chen — APPROVED** [review]

**Pull request merged**